File: Data Structure & Algorithm [Old]/Template/Binary Tree/Tree Restore/Construct Binary Tree from Inorder and Postorder Traversal/Construct Binary Tree from Inorder and Postorder Traversal.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recursive(inorder []int, postorder []int) *TreeNode {
	if len(postorder) <= 0 || len(inorder) <= 0 {
		return nil
	}

	currentVal := postorder[len(postorder)-1]
	currentIndex := 0
	for i, val := range inorder {
		if val == currentVal {
			currentIndex = i
			break
		}
	}

	node := &TreeNode{currentVal, nil, nil}

	leftpostorder := postorder[0:currentIndex]
	rightpostorder := postorder[currentIndex : len(postorder)-1]
	node.Right = recursive(inorder[currentIndex+1:], rightpostorder)
	node.Left = recursive(inorder[0:currentIndex], leftpostorder)

	return node
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	return recursive(inorder, postorder)
}

// test
func levelOrder(root *TreeNode) []int {
	res := []int{}
	queue := []*TreeNode{}

	if root == nil {
		return res
	}

	queue = append(queue, root)
	for len(queue) > 0 {
		size := len(queue)
		for i := 0; i < size; i++ {
			current := queue[0]
			queue = queue[1:]
			res = append(res, current.Val)
			if current.Left != nil {
				queue = append(queue, current.Left)
			}

			if current.Right != nil {
				queue = append(queue, current.Right)
			}
		}
	}

	return res
}

func main() {
	inorder := []int{9, 3, 15, 20, 7}
	postorder := []int{9, 15, 7, 20, 3}
	res := buildTree(inorder, postorder)
	printTree := levelOrder(res)
	fmt.Println(printTree)
}

File: Data Structure & Algorithm/Models/Tree/Binary Tree/Tree Restore/Construct Binary Tree from Inorder and Postorder Traversal/Construct Binary Tree from Inorder and Postorder Traversal.py
from typing import List, Optional


class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def buildTree(
        self,
        inorder: List[int],
        postorder: List[int],
    ) -> Optional[TreeNode]:

        def recursive(
            inorder: List[int],
            postorder: List[int],
        ) -> Optional[TreeNode]:
            if len(postorder) <= 0 or len(inorder) <= 0:
                return None

            currentVal = postorder.pop()
            currentIndex = inorder.index(currentVal)
            node = TreeNode(currentVal)

            node.right = recursive(inorder[currentIndex + 1:], postorder)
            node.left = recursive(inorder[0:currentIndex], postorder)

            return node

        return recursive(inorder, postorder)


# test
def levelOrder(root: TreeNode | None) -> List[int]:
    res = []
    queue = []

    if root == None:
        return res

    queue.append(root)
    while (len(queue) > 0):
        size = len(queue)
        for _ in range(size):
            current = queue.pop(0)
            res.append(current.val)
            if (current.left):
                queue.append(current.left)
            if (current.right):
                queue.append(current.right)
    return res


inorder = [9, 3, 15, 20, 7]
postorder = [9, 15, 7, 20, 3]
solution = Solution()
res = solution.buildTree(inorder, postorder)
printTree = levelOrder(res)
print(printTree)
